components/layout/display_list_builder.rs

```diff
@@ -1024,6 +1028,201 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
     }
 }
 
+/// Subtract offsets from a bounding box.
+///
+/// As an example if the bounds are the border-box and the border
+/// is provided as offsets the result will be the padding-box.
+fn calculate_inner_bounds(mut bounds: Rect<Au>, offsets: SideOffsets2D<Au>) -> Rect<Au> {
+    bounds.origin.x += offsets.left;
+    bounds.origin.y += offsets.top;
+    bounds.size.width -= offsets.horizontal();
+    bounds.size.height -= offsets.vertical();
+    bounds
+}
+
+
+/// For a given area and an image compute how big the
+/// image should be displayed on the background.
+fn compute_background_image_size(bg_size: BackgroundSize<LengthOrPercentageOrAuto>,
+                                 bounds_size: Size2D<Au>,
+                                 intrinsic_size: Option<Size2D<Au>>)
+                                 -> Size2D<Au> {
+    match intrinsic_size {
+        None => {
+            match bg_size {
```

Comment:
Well, I would've keep the early return, just using `match`, but it's not a big deal, this looks fine too, thanks!